python/ml4ir/applications/ranking/t_test.py

```diff
@@ -1,5 +1,324 @@
+    # Check the number of arguments that are None
+    n_none = sum([v is None for v in [d, n, power, alpha]])
+    if n_none != 1:
+        raise ValueError("Exactly one of n, d, power, and alpha must be None.")
+
+    # Safety checks
+    assert alternative in [
+        "two-sided",
+        "greater",
+        "less",
+    ], "Alternative must be one of 'two-sided' (default), 'greater' or 'less'."
+    assert contrast.lower() in ["one-sample", "paired", "two-samples"]
+    tsample = 2 if contrast.lower() == "two-samples" else 1
+    tside = 2 if alternative == "two-sided" else 1
+    if d is not None and tside == 2:
+        d = abs(d)
+    if alpha is not None:
+        assert 0 < alpha <= 1
+    if power is not None:
+        assert 0 < power <= 1
+
+    if alternative == "less":
+
+        def func(d, n, power, alpha):
+            dof = (n - 1) * tsample
+            nc = d * np.sqrt(n / tsample)
+            tcrit = stats.t.ppf(alpha / tside, dof)
+            return stats.nct.cdf(tcrit, dof, nc)
+
+    elif alternative == "two-sided":
+
+        def func(d, n, power, alpha):
+            dof = (n - 1) * tsample
+            nc = d * np.sqrt(n / tsample)
+            tcrit = stats.t.ppf(1 - alpha / tside, dof)
+            return stats.nct.sf(tcrit, dof, nc) + stats.nct.cdf(-tcrit, dof, nc)
+
+    else:  # Alternative = 'greater'
+
+        def func(d, n, power, alpha):
+            dof = (n - 1) * tsample
+            nc = d * np.sqrt(n / tsample)
+            tcrit = stats.t.ppf(1 - alpha / tside, dof)
+            return stats.nct.sf(tcrit, dof, nc)
+
+    # Evaluate missing variable
+    if power is None:
+        # Compute achieved power given d, n and alpha
+        return func(d, n, power=None, alpha=alpha)
+
+    elif n is None:
+        # Compute required sample size given d, power and alpha
+
+        def _eval_n(n, d, power, alpha):
+            return func(d, n, power, alpha) - power
+
+        try:
+            return brenth(_eval_n, 2 + 1e-10, 1e07, args=(d, power, alpha))
+        except ValueError:  # pragma: no cover
+            return np.nan
+
+    elif d is None:
+        # Compute achieved d given sample size, power and alpha level
+        if alternative == "two-sided":
+            b0, b1 = 1e-07, 10
+        elif alternative == "less":
+            b0, b1 = -10, 5
+        else:
+            b0, b1 = -5, 10
+
+        def _eval_d(d, n, power, alpha):
+            return func(d, n, power, alpha) - power
+
+        try:
+            return brenth(_eval_d, b0, b1, args=(n, power, alpha))
+        except ValueError:  # pragma: no cover
+            return np.nan
+
+    else:
+        # Compute achieved alpha (significance) level given d, n and power
+
+        def _eval_alpha(alpha, d, n, power):
+            return func(d, n, power, alpha) - power
+
+        try:
+            return brenth(_eval_alpha, 1e-10, 1 - 1e-10, args=(d, n, power))
+        except ValueError:  # pragma: no cover
+            return np.nan
```

Comment:
@mohazahran You can use [lambda functions](https://www.geeksforgeeks.org/python-lambda-anonymous-functions-filter-map-reduce/) to wrap this I think. (I still haven't looked at this in detail).

Reply:
Same as the previous comment.

Reply:
Yes, I made these comments before I realized this was from pinguoin. Haha